### ISSUE TITLE
'id' array in where clause for 'find' blueprint is handled as 'findOne'

### ISSUE THREAD

**Issue by arryon** [balderdashy/sails#2630]
I want to restrict the documents a user can retrieve for a certain model when invoking the 'find' blueprint action. Therefore, I wrote a middleware that retrieves id's, and puts them as an array in the 'where' object in req.options. According to the code, and the documentation, this should get me all documents from the database that matches any of these id values.

However, against expectation, the 'findOne' blueprint is triggered because of legacy compatibility with v0.9. In sails/lib/hooks/blueprints/actions/find.js the following is written from line 34:
```javascript
  // If an `id` param was specified, use the findOne blueprint action
  // to grab the particular instance with its primary key === the value
  // of the `id` param.   (mainly here for compatibility for 0.9, where
  // there was no separate `findOne` action)
  if ( actionUtil.parsePk(req) ) {
    return require('./findOne')(req,res);
  }
```

The actual method checks if there is an 'id' attribute in the where clause, correctly identifying it as the primary key, but the result is invocation of the findOne method, which will of course only return one document.

Right now I've disabled the code because it doesn't look like I need it. However, the documentation says nothing about this race condition. Here, http://sailsjs.org/#/documentation/concepts/ORM/Querylanguage.html, it is given that we can pass arrays in the where clause. the 'id' attribute shouldn't be treated any differently in that case. On the contrary, it most likely points to unique documents that we want to be able to retrieve using find.

**Comment by sgress454** [comment]
Hi @arryon, thanks for pointing this out.  Would be open to a PR that fixes this by checking whether the result of `parsePk` is an array!

**Comment by arryon** [comment]
Hi, I can easily submit a PR for this since it's a small fix. However, I'm struggling a bit to find the correct way of loading fixtures into a mock database, and invoking the find blueprint in the correct way in order to test my fix. It might take me some searching still.

**Comment by arryon** [comment]
@sgress454 am I correct when I notice that the integration tests from Sails won't help me here, since they are not up to date with the latest? I see a sample app with mock data I could use to test the find blueprint, but using which database connection? And there is no clearcut place to put my test: should it be in itegration, or in hooks?

**Comment by arryon** [comment]
@sgress454 See above for my fix. The existing tests are passed. If you want an additional test, I need some more help using the helper app from the tests to generate my use case. Waiting for feedback before I submit a PR.

**Comment by kevinob11** [comment]
I ran into this issue a few days ago and had to override the blueprint. This looks like a good solution to me.

**Comment by tjwebb** [comment]
@arryon can you go ahead and submit a PR with the fix? It'll be easier to review that way.

**Comment by sgress454** [comment]
Agreed--the fix looks good, go ahead an submit your PR along with a couple of tests in [router.specifiedRoutes.test.js](https://github.com/balderdashy/sails/blob/master/test/integration/router.specifiedRoutes.test.js)--one showing `req.options.where.id` set to an int and one showing it set to an array.  

Your tests will need a `before` which creates a couple of model instances (you can use `Empty.create()`), and creates a route using `httpHelper.writeRoutes` that has something like `options: {where: {id: [1,2]}}`. (See an example use of `writeRoutes` [here](https://github.com/balderdashy/sails/blob/master/test/integration/router.specifiedRoutes.test.js#L47)).  Then you can use `httpHelper.testRoute` to test it.

**Comment by arryon** [comment]
Understood, I'll try and make time tomorrow to integrate these tests and submit a PR

**Comment by CWyrtzen** [comment]
Any update here?

**Comment by sailsbot** [comment]
Thanks for posting, @arryon.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
